src/common/convenience.h

```diff
@@ -0,0 +1,142 @@
+// is nonnegative integer.
+//
+// The `c` is now the value that `__COUNTER__` expands to and we can
+// safely paste into a token.
+//
+#define GENSY_M_(p, c) p##c
+//
+// If we paste here, we'll get `PREFIX__COUNTER__`.
+//
+#define GENSYM_(p, c) GENSY_M_(p, c)
+//
+// Indirect to force `__COUNTER__` to be evaluated
+//
+#define GENSYM(prefix) GENSYM_(prefix, __COUNTER__)
+
+namespace ceph {
+
+// Lock Factories
+// ==============
+//
+// I used to, whenever I declared a mutex member variable of a class,
+// declare a pile of types like:
+// ```cpp
+// using unique_lock = ::std::unique_lock<decltype(membermutex);
+// ```
+// to avoid having to type that big, long type at every use. It also
+// let me change the mutex type later. It's inelegant and breaks down
+// if you have multiple mutices of differing types in the same
+// class. So here are some lock factories.
+template<typename Mutex, typename ...Args>
+auto uniquely_lock(Mutex&& m, Args&& ...args)
+  -> std::unique_lock<remove_reference_t<Mutex> > {
+  return std::unique_lock<remove_reference_t<Mutex> >(
+    std::forward<Mutex>(m), std::forward<Args>(args)... );
+}
+
+template<typename Mutex, typename ...Args>
+auto sharingly_lock(Mutex&& m, Args&& ...args)
+  -> boost::shared_lock<remove_reference_t<Mutex> > {
+  return
+    boost::shared_lock<remove_reference_t<Mutex> >(
+      std::forward<Mutex>(m), std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename ...Args>
+auto shuniquely_lock(std::unique_lock<Mutex>&& m, Args&& ...args)
+  -> shunique_lock<remove_reference_t<Mutex> > {
+  return shunique_lock<remove_reference_t<Mutex> >(
+    std::forward<std::unique_lock<Mutex> >(m), std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename ...Args>
+auto shuniquely_lock(boost::shared_lock<Mutex>&& m, Args&& ...args)
+  -> shunique_lock<remove_reference_t<Mutex> > {
+  return shunique_lock<remove_reference_t<Mutex> >(
+    std::forward<boost::shared_lock<Mutex> >(m),
+    std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename ...Args>
+auto shuniquely_lock(Mutex&& m, Args&& ...args)
+  -> shunique_lock<remove_reference_t<Mutex> > {
+  return shunique_lock<remove_reference_t<Mutex> >(
+    std::forward<Mutex>(m), std::forward<Args>(args)...);
+}
+}
+
+// Lock Guard
+// ----------
+//
+// I dislike C preprocessor macros, but a template won't do what I
+// want. Fundamentally, I can't pass a variable in to a template and
+// substitute in the variable's type and go from there. Even if I
+// could, that would just get me to guardedly_lock<M>(M) which isn't
+// ideal.
+//
+// Thus, I have no choice.
+//
+// Simply
+// ```cpp
+// GUARDEDLY_LOCK(some_mutex);
+// ```
+// and there you are. You don't need to know the name of the variable.
+// You can't do anything with a lock_guard except exit the stack frame
+// where it lives.
+//
+#define GUARDEDLY_LOCK(M)                                   \
+  ::std::lock_guard<::ceph::remove_reference_t<decltype(M)>> \
```

Comment:
I didn't think make_lock_guard would work, given that lock_guard isn't movable. I hadn't realized that extending the lifetime of the temporary when bound to the reference would work that way, thank you. I'll add that and the with_* wrappers.